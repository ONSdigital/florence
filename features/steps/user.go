package steps

import (
	"context"
	"github.com/chromedp/cdproto/cdp"
	"github.com/chromedp/cdproto/network"
	"github.com/chromedp/chromedp"
	"github.com/pkg/errors"
	"net/http"
	"time"
)

type User interface {
	signIn(username string) error
	isSignedIn() bool
	signOut() error
	resetUser(fakeApi *FakeApi, ctx context.Context)
	setChromeCtx(ctx context.Context)
	setFakeApi(fakeApi *FakeApi)
	setAuthCookies()
}

func GenerateCookie(name, value, domain string) *http.Cookie {
	if domain == "" {
		domain = "localhost"
	}
	return &http.Cookie{
		Name:     name,
		Value:    value,
		Path:     "/",
		Domain:   domain,
		HttpOnly: true,
		Secure:   true,
		MaxAge:   0,
		SameSite: http.SameSiteStrictMode,
	}
}

func SetCookies(cookies ...*http.Cookie) chromedp.Action {
	return chromedp.ActionFunc(func(ctx context.Context) error {
		expr := cdp.TimeSinceEpoch(time.Now().Add(180 * 24 * time.Hour))
		for _, cookie := range cookies {
			err := network.SetCookie(cookie.Name, cookie.Value).
				WithExpires(&expr).
				WithDomain(cookie.Domain).
				WithPath(cookie.Path).
				WithHTTPOnly(cookie.HttpOnly).
				WithSecure(cookie.Secure).
				Do(ctx)
			if err != nil {
				return errors.New("set cookie error")
			}
		}
		return nil
	})
}
