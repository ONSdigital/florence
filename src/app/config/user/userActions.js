import types from "./userConstants";

export const userLoggedIn = (email, userType, isAdmin) => {
    return {
        type: types.USER_LOGGED_IN,
        email,
        userType,
        isAdmin,
    };
};

export const userLoggedOut = () => {
    return {
        type: types.USER_LOGGED_OUT,
    };
};

export const startRefeshAndSession = (refresh_expiry_time, session_expiry_time) => {
    let payload = {
        sessionTimer: {
            active: true,
            expire: session_expiry_time,
        },
    };

    if (refresh_expiry_time) {
        payload = {
            ...payload,
            refreshTimer: {
                active: true,
                expire: refresh_expiry_time,
            },
        };
    }
    return {
        type: types.START_REFRESH_AND_SESSION,
        payload,
    };
};

export const endSession = () => {
    return {
        type: types.END_SESSION,
    };
};
