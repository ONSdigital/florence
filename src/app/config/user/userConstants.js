const USER_LOGGED_OUT = "USER_LOGGED_OUT";
const USER_LOGGED_IN = "USER_LOGGED_IN";
const START_REFRESH_AND_SESSION = "START_REFRESH_AND_SESSION";
const END_REFRESH_SESSION = "END_REFRESH_SESSION";
const END_SESSION_SESSION = "END_SESSION_SESSION";

export default {
    USER_LOGGED_OUT,
    USER_LOGGED_IN,
    START_REFRESH_AND_SESSION,
    END_REFRESH_SESSION,
    END_SESSION_SESSION,
};
