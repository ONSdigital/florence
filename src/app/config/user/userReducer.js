import types from "./userConstants";

const initialState = {
    isAuthenticated: false,
    email: "",
    userType: "",
    isAdmin: false,
    sessionTimer: {
        active: false,
        expire: "",
    },
    refreshTimer: {
        active: false,
        expire: "",
    },
};

const userReducer = (state = initialState, action) => {
    switch (action.type) {
        case types.USER_LOGGED_IN: {
            return {
                ...state,
                isAuthenticated: true,
                email: action.email,
                userType: action.userType,
                isAdmin: action.isAdmin,
            };
        }
        case types.USER_LOGGED_OUT: {
            return {
                ...state,
                isAuthenticated: false,
                email: null,
                userType: null,
                isAdmin: null,
            };
        }
        case types.START_REFRESH_AND_SESSION: {
            return {
                ...state,
                ...action.payload,
            };
        }
        case types.START_REFRESH_AND_SESSION: {
            return {
                ...state,
                ...action.payload,
            };
        }
        default: {
            return state;
        }
    }
};

export default userReducer;
