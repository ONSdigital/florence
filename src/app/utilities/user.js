import http from '../utilities/http';
import { store } from '../config/store';
import { userLoggedIn } from '../config/actions';

export default class user {

    static get(email) {
        return http.get(`/zebedee/users?email=${email}`)
            .then(response => {
                return response;
            }).catch(error => {
                console.error(`Error getting user details for ${email} \n${error}`);
            })
    }

    static getAll() {
        return http.get(`/zebedee/users`)
            .then(response => {
                return response;
            }).catch(error => {
                console.error(`Error getting all users \n${error}`)
            })
    }

    static getPermissions(email) {
        return http.get(`/zebedee/permission?email=${email}`)
            .then(response => {
                return response;
            }).catch(error => {
                console.error(`Error getting user permissions for ${email} \n${error}`);
        });
    }

    static setUserState(user) {
        const email = user.email;
        let userType = '';
        if (user.editor) {
            userType = 'EDITOR'
        } else {
            userType = 'DATA-VIS'
        }
        const isAdmin = !!user.admin;
        store.dispatch(userLoggedIn(email, userType, isAdmin));
        localStorage.setItem("loggedInAs", email);
    }

}