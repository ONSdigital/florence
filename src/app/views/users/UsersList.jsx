import React, { useEffect, useState, useCallback } from "react";
import PropTypes from "prop-types";
import { useInput } from "../../hooks/useInput";
import { Link } from "react-router";
import filter from "lodash/filter";
import SimpleSelectableList from "../../components/simple-selectable-list/SimpleSelectableList";
import BackButton from "../../components/back-button";
import Magnifier from "../../icons/Magnifier";
import clsx from "clsx";
import Loader from "../../components/loader/Loader";

const mapUsers = (users, rootPath) => {
    return users.map(user => ({
        ...user,
        title: `${user.forename} ${user.lastname}`,
        details: [user.email],
        url: `${rootPath}/users/${user.id}`,
    }));
};

const UsersList = props => {
    const { active, suspended, loading, rootPath, loadUsers } = props;
    const [showActiveUsers, setShowActiveUsers] = useState(true);
    const [search, setSearch] = useInput("");

    useEffect(() => {
        loadUsers();
    }, []);

    const users = showActiveUsers ? mapUsers(active, rootPath) : mapUsers(suspended, rootPath);
    const getFilteredUsers = useCallback(() => {
        const str = search.value.toLowerCase();
        return filter(users, user => (user.forename + " " + user.lastname).toLowerCase().includes(str) || user.email.toLowerCase().includes(str));
    }, [search.value]);

    return (
        <div className="grid grid--justify-space-around">
            <div className="grid__col-9">
                <div className="grid grid--align-baseline">
                    <div className="grid__col">
                        <h1>Users</h1>
                    </div>
                    <div className="grid__col">
                        <Link role="link" className="margin-left--1" href={`${rootPath}/users/create`}>
                            Create new user
                        </Link>
                    </div>
                </div>
                {loading ? (
                    <div className="grid grid--justify-center grid--align-center">
                        <Loader classNames="loader--dark loader--large" />
                    </div>
                ) : (
                    <>
                        <div className="grid grid--justify-space-between grid--align-center margin-bottom--1">
                            <div className="grid__col-md-4">
                                <div className="search__input-group ">
                                    <Magnifier classes="search__icon-magnifier" viewBox="0 0 28 28" />
                                    <label htmlFor="search" className="visually-hidden">
                                        Search users by name or email
                                    </label>
                                    <input role="search" name="search" placeholder="Search users by name" {...search} />
                                </div>
                            </div>
                            <div className="grid__col-md-4">
                                <div className="btn-group">
                                    <div className="form__label inline">Show</div>
                                    <label htmlFor="active" className="visually-hidden">
                                        Show active users
                                    </label>
                                    <button
                                        onClick={() => setShowActiveUsers(true)}
                                        id="active"
                                        aria-pressed={showActiveUsers}
                                        className={clsx("btn btn--margin-left", showActiveUsers ? "btn--primary" : "btn--link")}
                                    >
                                        Active users
                                    </button>
                                    <label htmlFor="suspended" className="visually-hidden">
                                        Show suspended users
                                    </label>{" "}
                                    <button
                                        onClick={() => setShowActiveUsers(false)}
                                        aria-pressed={showActiveUsers}
                                        id="suspended"
                                        className={clsx("btn btn--margin-left", !showActiveUsers ? "btn--primary" : "btn--link")}
                                    >
                                        Suspended
                                    </button>
                                </div>
                            </div>
                        </div>
                        {users ? (
                            <SimpleSelectableList rows={search.value ? getFilteredUsers() : users} />
                        ) : (
                            <p className="padding-top--1">Nothing to show.</p>
                        )}
                    </>
                )}
            </div>
        </div>
    );
};

UsersList.propTypes = {
    rootPath: PropTypes.string.isRequired,
    params: PropTypes.shape({
        userID: PropTypes.string,
    }).isRequired,
    loggedInUser: PropTypes.object,
    active: PropTypes.arrayOf(
        PropTypes.shape({
            forename: PropTypes.string,
            lastname: PropTypes.string,
            email: PropTypes.string,
            status: PropTypes.string,
            active: PropTypes.boolean,
            id: PropTypes.string,
            status_notes: PropTypes.string,
            url: PropTypes.string,
            title: PropTypes.string,
            details: PropTypes.array,
        })
    ),
    loadUsers: PropTypes.func.isRequired,
    loading: PropTypes.bool,
};

export default UsersList;
