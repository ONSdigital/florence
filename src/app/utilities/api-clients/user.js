import http from "../http";
import { store } from "../../config/store";
import { reset } from "../../config/actions";
import { userLoggedIn, userLoggedOut } from "../../config/user/userActions";
import cookies from "../cookies";
import notifications from "../notifications";
import log from "../logging/log";
import sessionManagement from "../sessionManagement";
import { errCodes as errorCodes, errCodes } from "../errorCodes";
import { removeAuthState, setAuthState } from "../auth";

export default class user {
    static get(email) {
        return http.get(`/zebedee/users?email=${email}`);
    }

    static getAll = async params => {
        let config = {};
        if (window.getEnv != null) {
            config = window.getEnv();
        }
        if (config.enableNewSignIn) {
            let queryString = "";
            Object.keys(params).map(key => {
                queryString = queryString ? `${queryString}&${key}=${params[key]}` : `?${key}=${params[key]}`;
            });
            try {
                return await http.get(`/users${queryString}`);
            } catch (error) {
                if (error.status) {
                    if (error.status >= 400 && error.status < 500) {
                        switch (error.status) {
                            case 400: {
                                const notification = {
                                    type: "warning",
                                    message: errCodes.GET_USERS_UNEXPECTED_FILTER_ERROR,
                                    autoDismiss: 5000,
                                };
                                notifications.add(notification);
                                break;
                            }
                            case 404: {
                                const notification = {
                                    type: "warning",
                                    message: errCodes.GET_USERS_NOT_FOUND,
                                    autoDismiss: 5000,
                                };
                                notifications.add(notification);
                                break;
                            }
                            default: {
                                const notification = {
                                    type: "warning",
                                    message: errorCodes.GET_USERS_UNEXPECTED_ERROR_SHORT,
                                    isDismissable: true,
                                };
                                notifications.add(notification);
                                break;
                            }
                        }
                    } else {
                        const notification = {
                            type: "warning",
                            message: errCodes.GET_USERS_UNEXPECTED_ERROR_SHORT,
                            isDismissable: true,
                        };
                        notifications.add(notification);
                    }
                } else {
                    const notification = {
                        type: "warning",
                        message: errCodes.GET_USERS_UNEXPECTED_ERROR_SHORT,
                        isDismissable: true,
                    };
                    notifications.add(notification);
                }
                return error;
            }
        } else {
            return http.get(`/zebedee/users`);
        }
    };

    static create(body) {
        return http.post(`/zebedee/users`, body);
    }

    // TODO: new auth work
    static getUsers() {
        return http.get(`/users`);
    }

    // TODO: new auth work
    static createNewUser(body) {
        return http.post(`/users`, body);
    }
    // TODO: new auth work
    static getUser(id) {
        return http.get(`/users/${id}`);
    }

    // TODO: new auth work
    static updateUser(id, body) {
        return http.put(`/users/${id}`, body);
    }

    static getUserGroups(id) {
        return http.get(`/users/${id}/groups`);
    }

    static remove(email) {
        return http.delete(`/zebedee/users?email=" + ${email}`);
    }

    static setPassword(body) {
        return http.post(`/zebedee/password`, body);
    }

    static setPermissions(body) {
        return http.post(`/zebedee/permission`, body);
    }

    /**
     * Gets the permissions of the user specified by the email provided.
     *
     * @returns the user permissions response
     * @deprecated as part of migration to JWT sessions and will be removed soon. Use getPermissions() instead.
     */
    static getPermissionsForUser(email) {
        console.warn(
            "WARNING! Deprecated function called. Function 'getPermissionsForUser(email)' is deprecated, please use the new 'getPermissions()' instead"
        );
        return http.get(`/zebedee/permission?email=${email}`, true, true);
    }

    /**
     * Gets the permissions of the current authenticated user (based on their access token)
     *
     * @returns the user permissions response
     */
    static getPermissions() {
        return http.get(`/zebedee/permission`, true, true);
    }

    static signIn(body) {
        return http.post("/tokens", body, true, true, true);
    }

    // TODO: new auth work
    static deleteTokens() {
        return http.delete("/tokens");
    }

    static setForgottenPassword(body) {
        return http.put("/users/self/password", body, true, true);
    }

    static renewSession(body) {
        return http.put("/tokens/self", body, true, false);
    }

    static expireSession() {
        return http.delete("/tokens/self", true, true);
    }

    // This is a temporary fix for 925: Login fails after going from new login to old login
    static deleteCookies() {
        return http.delete("/cookies");
    }

    static getOldUserType(user) {
        // TAKEN FROM OLD FLORENCE
        if (user.admin) {
            return "PUBLISHING_SUPPORT";
        } else if (user.editor && !user.dataVisPublisher) {
            return "PUBLISHING_SUPPORT";
        } else if (user.editor && user.dataVisPublisher) {
            return "DATA_VISUALISATION";
        } else if (!user.admin && !user.editor && !user.dataVisPublisher) {
            return "VIEWER";
        }
    }

    static getUserRole(isAdmin, isEditor) {
        let role = "";
        if (isEditor) {
            role = "EDITOR";
        }
        if (isAdmin) {
            role = "ADMIN";
        }
        if (!isEditor && !isAdmin) {
            role = "VIEWER";
        }
        return role;
    }

    static setUserState(user) {
        setAuthState(user);
        const email = user.email;
        const role = this.getUserRole(user.admin, user.editor);
        const isAdmin = !!user.admin;
        store.dispatch(userLoggedIn(email, role, isAdmin));
    }

    static logOut() {
        function clearCookies() {
            const accessTokenCookieRemoved = cookies.remove("access_token");
            if (!accessTokenCookieRemoved) {
                if (!config.enableNewSignIn) {
                    user.deleteCookies()
                        .then(function () {
                            console.debug("[FLORENCE] Deleted HTTP Cookies");
                        })
                        .catch(err => console.error(err));
                }
                console.warn(`Error trying to remove 'access_token' cookie`);
            }
            if (cookies.get("collection")) {
                cookies.remove("collection");
            }
            if (cookies.get("id_token")) {
                cookies.remove("id_token");
            }
            removeAuthState();
            store.dispatch(userLoggedOut());
            store.dispatch(reset());
        }

        const config = window.getEnv();
        if (config.enableNewSignIn) {
            user.expireSession()
                .catch(error => {
                    if (error.status === 400) {
                        const notification = {
                            type: "warning",
                            message: "An error occurred during sign out 'InvalidToken', please contact a system administrator",
                            isDismissable: true,
                            autoDismiss: 20000,
                        };
                        notifications.add(notification);
                        console.error("Error occurred sending DELETE to /tokens/self - InvalidToken");
                        log.event("error on sign out sending delete to /tokens/self failed with an invalid token", log.error(error));
                    } else {
                        const notification = {
                            type: "warning",
                            message: "Unexpected error occurred during sign out",
                            isDismissable: true,
                            autoDismiss: 20000,
                        };
                        notifications.add(notification);
                        console.error("Error occurred sending DELETE to /tokens/self");
                        log.event("error on sign out sending delete to /tokens/self failed with an unexpected error", log.error(error));
                    }
                    clearCookies();
                })
                .finally(() => {
                    clearCookies();
                    sessionManagement.removeTimers();
                });
        } else {
            clearCookies();
        }
    }

    static updatePassword(body) {
        return http.post("/zebedee/password", body, undefined, true);
    }
}
