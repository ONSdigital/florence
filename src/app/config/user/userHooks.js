import user from "../../utilities/api-clients/user";
import { useEffect, useState } from "react";
import { startRefeshAndSession } from "../../config/user/userActions";
import sessionManagement from "../../utilities/sessionManagement";
import { getAuthState } from "../../utilities/auth";
import fp from "lodash/fp";

const nonAuthRoutes = ["/florence/login", "/florence/forgotten-password", "/florence/password-reset"];

export const useGetPermissions = (authState, setShouldUpdateAccessToken) => {
    const [userState, setUserState] = useState();
    useEffect(() => {
        if (!authState) {
            user.getPermissions().then(userData => {
                // TODO this needs to be removed when we get a correct 401 status back from zebedee.
                if (userData === "Access Token required but none provided.") {
                    console.error("Error fetching permissions");
                    return null;
                }
                setShouldUpdateAccessToken(true);
                setUserState(userData);
            });
        } else {
            setShouldUpdateAccessToken(false);
            setUserState(authState);
        }
    }, []);
    return userState;
};

export const useUpdateTimers = (props, sessionTimerIsActive, dispatch) => {
    useEffect(() => {
        if (!nonAuthRoutes.includes(props.location.pathname) && !sessionTimerIsActive) {
            const enableNewSignIn = fp.get("config.enableNewSignIn")(props);
            const authState = getAuthState(); // Get the lastest authState
            const session_expiry_time = fp.get("session_expiry_time")(authState);
            const refresh_expiry_time = new Date(fp.get("refresh_expiry_time")(authState));
            if (sessionManagement.isSessionExpired(session_expiry_time)) {
                if (enableNewSignIn) {
                    console.debug("Timers / enableNewSignIn: requesting a new access_token");
                    user.renewSession()
                        .then(res => {
                            // update the authState, start the session timer with the nest session response value
                            // & restart the refresh timer with the existing refresh value.
                            const expirationTime = sessionManagement.convertUTCToJSDate(fp.get("expirationTime")(res));
                            sessionManagement.initialiseSessionExpiryTimers(expirationTime, refresh_expiry_time);
                            dispatch(startRefeshAndSession(fp.get("refresh_expiry_time")(authState), expirationTime));
                        })
                        .catch(err => console.error(err));
                } else {
                    console.debug("[FLORENCE] Timers: extending session & refresh timers");
                    // If we are not behind the enableNewSignIn then just extend the session & refresh for 12 hours
                    const expireTimes = sessionManagement.createDefaultExpireTimes(12);
                    sessionManagement.initialiseSessionExpiryTimers(expireTimes.session_expiry_time, expireTimes.refresh_expiry_time);
                }
            } else {
                if (enableNewSignIn) {
                    console.debug("[FLORENCE] Timers: starting timers");
                    // The user has refreshed the page but the session is not expired, so just restart the timers
                    // for both refresh & session.
                    const session_expiry_time = fp.get("session_expiry_time")(authState);
                    sessionManagement.initialiseSessionExpiryTimers(new Date(session_expiry_time), refresh_expiry_time);
                    dispatch(startRefeshAndSession(fp.get("refresh_expiry_time")(authState), session_expiry_time));
                } else {
                    console.debug("[FLORENCE] Timers: starting timers");
                    // The user has refreshed the page but the session is not expired, so just restart the timers
                    // for both refresh & session.
                    const session_expiry_time = fp.get("session_expiry_time")(authState);
                    sessionManagement.initialiseSessionExpiryTimers(new Date(session_expiry_time), null);
                    dispatch(startRefeshAndSession(null, session_expiry_time));
                }
            }
        }
    }, [sessionTimerIsActive]);
};
