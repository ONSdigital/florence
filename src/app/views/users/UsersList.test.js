import React from "react";
import { UsersList, getAllUsers, mapUserToState } from "./UsersList";
import user from "../../utilities/api-clients/user";
import renderer from "react-test-renderer";
import { mount } from "enzyme";

console.error = () => {};

const withTimeout = (done, fn) => {
    const timeoutId = setTimeout(() => {
        fn();
        clearTimeout(timeoutId);
        done();
    });
};

jest.mock("../../utilities/logging/log", () => {
    return {
        event: jest.fn(() => {}),
        data: jest.fn(() => {}),
        error: jest.fn(() => {}),
    };
});

jest.mock("../../utilities/notifications", () => {
    return {
        add: jest.fn(() => {}),
        remove: () => {},
    };
});

jest.mock("../../utilities/api-clients/user", () => {
    return {
        getAll: jest.fn(() => {
            return Promise.resolve(mockedAllUsers);
        }),
    };
});

jest.mock("../../utilities/auth", () => {
    return {
        isAdmin: jest.fn(() => {
            return true;
        }),
    };
});

const mockedAllUsers = {
    count: 45,
    users: [
        {
            forename: "Test",
            lastname: "user",
            email: "test@test.com",
            groups: [],
            status: "CONFIRMED",
            active: true,
            id: "aaa-bbb-ccc",
            status_notes: "",
        },
        {
            forename: "Test",
            lastname: "user 2",
            email: "test2@test.com",
            groups: [],
            status: "CONFIRMED",
            active: true,
            id: "ddd-eee-fff",
            status_notes: "",
        },
        {
            forename: "Test",
            lastname: "user 3",
            email: "test3@test.com",
            groups: [],
            status: "CONFIRMED",
            active: true,
            id: "ggg-hhh-iii",
            status_notes: "",
        },
    ],
};

let dispatchedActions = [];

const mockDispatch = event => {
    dispatchedActions.push(event);
};
const rootPath = "/florence";

const defaultProps = {
    dispatch: mockDispatch,
    rootPath: rootPath,
    params: {
        userID: "",
    },
};

describe("Calling getAllUsers function", () => {
    it("fetches data for all users", done =>
        withTimeout(done, () => {
            const getUserCalls = user.getAll.mock.calls.length;
            getAllUsers(mockDispatch, rootPath, () => {});
            expect(user.getAll.mock.calls.length).toBe(getUserCalls + 1);
        }));

    it("adds all users to state", () => {
        getAllUsers(mockDispatch, rootPath, () => {});
        expect(dispatchedActions[0].type).toBe("ADD_ALL_USERS");
        expect(dispatchedActions[0].users.length).toBe(mockedAllUsers.users.length);
    });
});

describe("Mapping users to state", () => {
    it("maps correctly", () => {
        const expectedValue = {
            ...mockedAllUsers.users[0],
            id: mockedAllUsers.users[0].email,
            details: [mockedAllUsers.users[0].email],
            title: `${mockedAllUsers.users[0].forename} ${mockedAllUsers.users[0].lastname}`,
            url: `/florence/users/${mockedAllUsers.users[0].email}`,
        };
        const returnValue = mapUserToState("/florence", mockedAllUsers.users[0]);
        expect(returnValue).toMatchObject(expectedValue);
    });
});

describe("UsersList", () => {
    const mappedUser1 = mapUserToState("/florence", mockedAllUsers.users[0]);
    const mappedUser2 = mapUserToState("/florence", mockedAllUsers.users[1]);
    const defaultProps = {
        dispatch: mockDispatch,
        rootPath: rootPath,
        params: {
            userID: "",
        },
        users: [mappedUser1, mappedUser2],
    };

    const component = mount(<UsersList {...defaultProps} />);
    it("should display a simple selectable list", () => {
        expect(component.find(".simple-select-list").length).toBe(1);
        expect(component.find(".simple-select-list__item").length).toBe(2);
    });

    it("matches the snapshot", () => {
        const componentForSnapshot = renderer.create(<UsersList {...defaultProps} />);
        expect(componentForSnapshot.toJSON()).toMatchSnapshot();
    });
});
