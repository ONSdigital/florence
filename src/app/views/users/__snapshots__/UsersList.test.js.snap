// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UsersList matches the snapshot 1`] = `
<div
  className="grid grid--justify-space-around"
>
  <div
    className="grid__col-9"
  >
    <div
      className="margin-top--2"
    >
      ◀ 
      <a
        onClick={[Function]}
        style={Object {}}
      >
        Back
      </a>
    </div>
    <div
      className="grid grid--align-baseline"
    >
      <div
        className="grid__col-1"
      >
        <h1>
          Users
        </h1>
      </div>
      <div
        className="grid__col-3"
      >
        <a
          className="margin-left--1"
          href="/florence/users/create"
          onClick={[Function]}
          style={Object {}}
        >
          Create new user
        </a>
      </div>
    </div>
    <div
      className="grid"
    >
      <div
        className="grid__col-4"
      >
        <div
          className="form__input"
        >
          <label
            className="form__label"
            htmlFor="search-content-types"
          />
          <input
            autoComplete="new-password"
            autoFocus={false}
            className="input"
            disabled={false}
            id="search-content-types"
            name="search-content-types"
            onChange={[Function]}
            onFocus={[Function]}
            placeholder="Search user by name or email"
            type="text"
          />
        </div>
      </div>
    </div>
    <ul
      className="list list--neutral simple-select-list"
    >
      <li
        className="simple-select-list__item"
      >
        <div
          className="simple-select-list__col simple-select-list__cols-1"
        >
          <a
            onClick={[Function]}
            style={Object {}}
          >
            <p
              className="simple-select-list__title"
            >
              Test user
            </p>
          </a>
          <p>
            test@test.com
          </p>
        </div>
      </li>
      <li
        className="simple-select-list__item"
      >
        <div
          className="simple-select-list__col simple-select-list__cols-1"
        >
          <a
            onClick={[Function]}
            style={Object {}}
          >
            <p
              className="simple-select-list__title"
            >
              Test user 2
            </p>
          </a>
          <p>
            test2@test.com
          </p>
        </div>
      </li>
      
    </ul>
  </div>
</div>
`;
